ci: Automate Dependabot configuration generation

vdemeester opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

- Generate the dependabot configuration to handle LTS branches in
  order to prevent release branches dependencies drift and reduce manual
  effort in maintaining per-branch dependency updates
- For release branches, only bump dependencies on patch versions
- Enable weekly automated checks with PR creation on changes

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

/kind misc
/area automation

Closes #8572
<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

afrittoli reviewed (commented):
Thanks for this. Perhaps for some ecosystems (like GHA), we could setup auto merge if CI passes on release branches, otherwise it will become a lot of work only to approve all the PRs for the various branches 😅

waveywaves reviewed (approved):
/lgtm

anticipating this to be opened on Monday !

aThorp96 commented on .github/dependabot.config.yml:
```diff
@@ -0,0 +1,125 @@
+# List of release branches to monitor (in addition to main branch)
+# These should be active LTS release branches
+# Check releases.md for the current list of supported releases
+release-branches:
+  - release-v1.9.x
+  - release-v1.6.x
+  - release-v1.3.x
+  - release-v1.0.x
```
This should be be automatically updated I think, it should be coming from a source of truth elsewhere

vdemeester replied:
Right, for now, the only "source of truth" is the release.md markdown, but not necessarily easy to parse. For now I thought it would make sense to have a post release checklist item for this.. but agreeing that mid/long term it needs to be automated.

aThorp96 replied:
Hmmm, yeah that's not super clean to parse as the source of truth. It could be done using an `awk` oneliner but it's a little ugly/archaic so not a blocker

E.g. this awk script searches for any 3rd-level heading under `## Release` (stops searching at the next 2nd-level heading) which match the regex `### \d+\.\d+`, then rewrites the `\d\.\d` version into the branch name. 
```bash
 $ awk '/^## Release$/{found=1; next} found && /^## /{exit} found && /^### v[0-9]+\.[0-9]+/{sub(/^### /, "release-"); sub(/( \(LTS\))?\s*$/, ".x"); print}' releases.md
release-v1.11.x
release-v1.9.x
release-v1.6.x
release-v1.3.x
release-v1.0.x
```
Note though that regex doesn't check for ` \(LTS\)$` so it lists `release-v1.11.x` which is not included in this PR's list. 


If we want to leave this list as hardcoded can we update the post-release checklist in this same PR?

aThorp96 commented on hack/generate-dependabot.sh:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+# Copyright 2025 The Tekton Authors
```
```suggestion
# Copyright 2026 The Tekton Authors
```

vdemeester replied:
Ahhah yeah because I created this PR in 2025 😛

aThorp96 commented on hack/generate-dependabot.go:
```diff
@@ -0,0 +1,169 @@
+// Copyright 2025 The Tekton Authors
```
```suggestion
// Copyright 2026 The Tekton Authors
```

aThorp96 commented on hack/generate-dependabot.sh:
```diff
@@ -0,0 +1,33 @@
+
+REPO_ROOT=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+cd "${REPO_ROOT}"
+
```
Something like this would work to generate the release branch list (currently only selects LTS release branches)
```suggestion
echo "Updating .github/dependabot.config.yml

echo "Generating release branch list from release.md active releases
yq -i '.release-branches = []' .github/dependabot.config.yml
awk '/^## Release$/{found=1; next} found && /^## /{exit} found && /^### v[0-9]+\.[0-9]+ \(LTS\)/{sub(/^### /, "release-"); sub(/ \(LTS\)\s*$/, ".x"); print}' releases.md | 
while IFS= read -r branch; do
    yq -i ".release-branches += ${branch}" .github/dependabot.config.yml
done

```

vdemeester replied:
Alright, let's try this approach 😬

vdemeester committed:
ci: Automate Dependabot configuration generation

- Generate the dependabot configuration to handle LTS branches in
  order to prevent release branches dependencies drift and reduce manual
  effort in maintaining per-branch dependency updates
- For release branches, only bump dependencies on patch versions
- Enable weekly automated checks with PR creation on changes
- Generate release-branches list dynamically from releases.md

Closes #8572

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

vdemeester committed:
fix: address zizmor findings in dependabot-regen workflow

- Add persist-credentials: false to checkout step
- Move all ${{ }} expressions from run: blocks to env: blocks
  to prevent template injection vulnerabilities

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

vdemeester commented:
/retest

afrittoli commented:
/lgtm
